### ISSUE TITLE
Create a `set_storage_or_clear` that deletes an storage item if 0u256 is passed as value

### ISSUE THREAD

**Issue by athei** [paritytech/polkadot-sdk#6944]
On Ethereum writing zero to a key constitutes deleting the key.  We should add a new host function that behaves the same as `set_storage` except that the value is deleted if zero is passed. This is useful because it allows refunding the storage deposit when a key is deleted.

Note that revive could do that without any new host function by calling `clear_storage` in case the key is zero. However, this would increase contract code by adding this additional check.

In the same vein we could also add a `get_storage_or_zero` that returns 0 of the key does not exist. It would remove some code from the contract that is needed to check the return value of `get_storage`.

## Todo

We should add those two new host functions (i.e as new functions [here](https://github.com/paritytech/polkadot-sdk/blob/42e9de7f4fdb191cd531b443c68cfe19886ec311/substrate/frame/revive/src/wasm/runtime.rs#L1159)).

- [ ] Add `set_storage_or_clear` host function (same semantic as [`SSTORE`](https://www.evm.codes/?fork=cancun#55) + clearing behaviour above)
- [ ] Add `get_storage_or_zero` host function (same semantic as [`SLOAD`](https://www.evm.codes/?fork=cancun#54)).

Both will use hardcoded fixed size keys and fixed size output. Both 256bit. As opposed to variable length keys and values as the existing host functions.

cc @xermicus

**Comment by xermicus** [comment]
Good idea, I think we might just want to provide EVM tailored storage setters and getters but leave the more generic ones intact for other languages that are more flexible?

Regarding `set_storage`, other contract languages might not be limited to 32 byte storage values, and if the value is large the check might get costly and unwanted? Maybe for Solidity we should just add a `set_storage_or_delete` which accepts a single 32 byte key and value and deletes the key if the value equals zero?

For the getter, how I implemented it in `revive` for now is contracts zero out the output buffer of `get_storage` and ignore the return value. Which results in the desired behavior. However it would be slightly more efficient if the pallet would do it. We could just change `get_storage` to resemble the EVM semantics, however other languages might have a use case for a fallible `get_storage` method, so adding an infallible `get_storage_or_zero` host function seems sensible to me.

**Comment by athei** [comment]
Yes the idea is two add two new functions tailored towards EVM and leave the other ones intact.

**Comment by xermicus** [comment]
Sounds good. I can take care of this, i.e. implementing those pallet-revive plus a companion in revive. But I'll not assign it just yet as it's also a good first issue for anyone wanting to help out.

**Comment by athei** [comment]
Added the information regarding the getter to the initial description.
